### DISCUSSION TITLE
[Component] Table of Contents (with proper highlighting)

### DISCUSSION THREAD

**Discussion by goustaff** [shadcn-ui/ui#2990] [Ideas]
Hello!
Not long ago, I coded a table of contents (ToC) to play with some blog pages.

At first, it didn't look complicated. But, in practice, I found it extremely trickier than expected.
The main problem is highlighting (aka "HL").

See it by yourself with some real world examples:
- https://blowfish.page/docs/installation/
Scroll down until the highlight reaches "Download the Blowfish theme", then, continue to scroll down...
The HL will be like "ghosted" in a wide dead zone.

- https://thedevdavid-digital-garden.vercel.app/posts/choosing-providers
Click on "Other analytics providers". The page scrolls until the section, but the HL will be triggered on "Other newsletter providers".

- https://biomejs.dev/linter/rules/#suspicious
Go on the "Suspicious" section. Then, scroll up until you're actually in the "Style" section (the "Suspicious" title isn't visible anymore at this point of scroll Y). The HL is stale, and still on "Suspicious", because it waits a title to be visible to update.

- https://biomejs.dev/linter/rules/
Hold Space key. The HL freezes.

- https://biomejs.dev/linter/rules/
Scroll until no section title is visible in your viewport. Refresh the page. You will be at the same scroll Y, and the HL will not be triggered. You're in the void!

---

I didn't understand why it's so difficult to get a decent HL on a simple table of contents.
And when I tried to implement it myself, I ran into all kinds of problems.

That's a shame, because it's very confusing for a user to have a highlight that bugs a lot on a table of contents component, especially if it's a documentation page.

I think it would be nice to add a component to shadcn/ui to solve this problem.

---

I started coding something on my own. It's probably not perfect, but I've done my best, and I'd be curious to start a discussion about this pain point.

https://github.com/gustaveWPM/dashboard_rtm/blob/main/src/components/ui/blog/BlogPostTocDesktopInner.tsx

Demo:
https://dashboard-rtm.vercel.app/en/blog/patch-notes/dashboard/post-06

**Comment by StashOfCode** [comment]
Could someone with admin rights remove this page? Barely any link mentioned is working, shows up on google search wasting everyone's time.

**Reply by goustaff** [reply]
Hello.
Yeah, I changed my repo visibility some days ago.
Sorry if it made you upset.

Were you searching for some code to create a ToC?
I don't think mine is the best to read, but I still ask.

**Reply by StashOfCode** [reply]
Hello Gustave,
No need to apologize - it's all right. I was a bit disappointed at first, but I quickly adjusted and adapted another ToC to fit my use case, and it worked perfectly. I appreciate your response - take care!

**Reply by goustaff** [reply]
Hello StashOfCode.
Would be glad to know which ToC you chose to adapt. :)

Regards.

**Reply by StashOfCode** [reply]
@gustaveWPM I copied [Mantine's ToC](https://mantine.dev/core/table-of-contents/) source code, then asked Claude Sonnet to simplify it by removing everything except for a few features I needed. Finally, I fixed whatever the model got wrong and wrote some css. So far so good.
